flake.nix

```diff
@@ -2,70 +2,88 @@
   description = "Weeder Nix";
 
   inputs = {
+    flake-utils.url = "github:numtide/flake-utils";
     nixpkgs.url = "github:NixOS/nixpkgs?ref=nixos-23.11";
     pre-commit-hooks.url = "github:cachix/pre-commit-hooks.nix";
+    systems.url = "github:nix-systems/x86_64-linux";
   };
 
-  outputs = { self, nixpkgs, pre-commit-hooks }:
+  outputs = { self, flake-utils, nixpkgs, pre-commit-hooks, systems }:
     let
-      system = "x86_64-linux";
-      pkgs = import nixpkgs {
-        inherit system;
-        overlays = [
-          (import ./nix/overlay.nix)
-        ];
+      lib = {
+        haskell = import ./nix/lib/haskell;
+        weeder = import ./nix/lib/weeder;
       };
+
     in
-    with pkgs.lib;
     {
-      overlays.${system} = import ./nix/overlay.nix;
-      lib.${system} = pkgs.weeder-nix;
-      checks.${system} = {
-        validity = self.lib.${system}.makeWeederCheck {
-          name = "validity";
-          reportOnly = true;
-          packages = [
-            "validity"
-            # "genvalidity"
-          ];
-        };
-        yesod = self.lib.${system}.makeWeederCheck {
-          name = "yesod-weeder";
-          reportOnly = true;
-          packages = [
-            "yesod"
-            "yesod-auth"
-            "yesod-auth-oauth"
-            "yesod-bin"
-            "yesod-core"
-            "yesod-eventsource"
-            "yesod-form"
-            "yesod-form-multi"
-            "yesod-newsfeed"
-            "yesod-persistent"
-            "yesod-sitemap"
-            "yesod-static"
-            "yesod-test"
-            "yesod-websockets"
-          ];
-        };
-        pre-commit = pre-commit-hooks.lib.${system}.run {
-          src = ./.;
-          hooks = {
-            nixpkgs-fmt.enable = true;
-            deadnix.enable = true;
-            tagref.enable = true;
+      lib = lib.haskell // lib.weeder;
+
+      overlays.default = final: prev: {
+        haskell = prev.haskell // {
+          lib = prev.haskell.lib
+          // builtins.mapAttrs (_k: v: v final) lib.haskell
+          //
+          {
+            weeder = builtins.mapAttrs (_k: v: v final) lib.weeder;
           };
         };
       };
-      devShells.${system}.default = pkgs.mkShell {
-        name = "weeder-nix-shell";
-        buildInputs = with pre-commit-hooks.packages.${system};          [
-          pkgs.haskellPackages.weeder
-          nixpkgs-fmt
-          statix
-        ];
-        shellHook = self.checks.${system}.pre-commit.shellHook;
-      };
-    };
+    }
+    // flake-utils.lib.eachSystem (import systems) (system:
```

Comment:
Could we use this instead of yet another `flake-utils`?
```
    forAllSystems = function:
      nixpkgs.lib.genAttrs [
        "x86_64-linux"
        "aarch64-linux"
      ] (system: function nixpkgs.legacyPackages.${system});
```